File: 3097 Shortest Subarray With OR at Least K II/3097.py
class Solution:
    def minimumSubarrayLength(self, nums: List[int], k: int) -> int:
        BIT_LEN = 32
        bits = [0] * BIT_LEN

        def add(num):
            for i in range(BIT_LEN):
                if num & 1 << i:
                    bits[i] += 1
        
        def remove(num, cur):
            for i in range(BIT_LEN):
                b = 1 << i
                if num & b:
                    bits[i] -= 1
    
                    if bits[i] == 0:
                        cur -= b

            return cur
                

        n = len(nums)
        l = 0
        cur = 0
        
        INF = float('inf')
        ans = INF

        for r in range(n):
            num = nums[r]
            cur |= num
            add(num)

            while cur >= k and l <= r:
                ans = min(ans, r - l + 1)
                cur = remove(nums[l], cur)
                l += 1

        return -1 if ans == INF else ans

File: 3097 Shortest Subarray With OR at Least K II/3097.cpp
class Solution {
    int BIT_LEN = 32;
public:
    void add(int& num, vector<int>& bits, int& cur){
        for (int i = 0; i < BIT_LEN; i++){
            if (num & 1 << i){
                bits[i]++;
            }
        }
        cur |= num;
    }

    void remove(int& num, vector<int>& bits, int& cur){
        for (int i = 0; i < BIT_LEN; i++){
            int b = 1 << i;
            if (num & b){
                bits[i]--;
                if (bits[i] == 0){
                    cur -= b;
                }
            }
        }
    }

    int minimumSubarrayLength(vector<int>& nums, int k) {
        vector<int> bits(BIT_LEN, 0);
        int l = 0;
        int cur = 0;
        int ans = INT_MAX;

        for (int r = 0; r < nums.size(); r++){
            add(nums[r], bits, cur);

            while (cur >= k && l <= r){
                ans = min(ans, r - l + 1);
                remove(nums[l++], bits, cur);
            }
        }

        return ans == INT_MAX ? -1 : ans;
    }       
};
